interop-cats-effect/shared/src/main/scala/scalaz/zio/interop/bio/Concurrent2.scala

```diff
@@ -0,0 +1,334 @@
+    }
+  }
+
+  /**
+   * Returns an effect that executes both `fa1` and `fa2` in parallel
+   * and combines their results with `f`. If one of the two fails
+   * then the other will be interrupted.
+   *
+   * TODO: Example:
+   * {{{
+   *
+   * }}}
+   *
+   */
+  @inline def zipWithPar[E, A, B, C](fa1: F[E, A], fa2: F[E, B])(f: (A, B) => C)(
+    implicit
+    CD: ConcurrentData2[F],
+    MD: Monoid[E]
+  ): F[E, C] = {
+
+    def coordinate[AA, BB](f: (AA, BB) => C)(winner: Option[Either[E, AA]], loser: Fiber2[F, E, BB]): F[E, C] =
+      winner match {
+        case Some(Right(a)) => monad.map(loser.join)(f(a, _))
+        case Some(Left(winnerError)) =>
+          monad.flatMap(loser.cancel) {
+            case Some(Right(_))         => raiseError(winnerError)
+            case Some(Left(loserError)) => raiseError(MD.combine(loserError, winnerError))
+            case None                   => raiseError(MD.empty) // TODO: What to do in this case (None) ?
+          }
+        case None => raiseError(MD.empty) // TODO: What to do in this case (None) ?
+      }
+
+    val g = (b: B, a: A) => f(a, b)
+
+    raceWith(fa1, fa2)(coordinate(f), coordinate(g))
+  }
+
+  /**
+   * Returns an effect that executes both `fa1` and `fa2` in parallel
+   * and combines their results in a tuple. If one of the two fails
+   * then the other will be interrupted.
+   *
+   * TODO: Example:
+   * {{{
+   *
+   * }}}
+   *
+   */
+  @inline def zipPar[E, EE >: E, A, B](fa1: F[E, A], fa2: F[EE, B])(
+    implicit
+    CD: ConcurrentData2[F],
+    MD: Monoid[EE]
+  ): F[EE, (A, B)] =
+    zipWithPar(fa1, fa2)((a, b) => (a, b))
+
+  /**
+   * Returns an effect that executes both `fa1` and `fa2` in parallel
+   * and returns the result of `fa1`. If one of the two fails
+   * then the other will be interrupted.
+   *
+   * TODO: Example:
+   * {{{
+   *
+   * }}}
+   *
+   */
+  @inline def zipParLeft[E, EE >: E, A, B](fa1: F[E, A], fa2: F[EE, B])(
+    implicit
+    CD: ConcurrentData2[F],
+    MD: Monoid[EE]
+  ): F[EE, A] =
+    zipWithPar(fa1, fa2)((a, _) => a)
+
+  /**
+   * Returns an effect that races `fa` with all the effects in `xs`.
+   * The semantic is the same as `race` applied to a collection of
+   * parallel effects.
+   *
+   * TODO: Example:
+   * {{{
+   *
+   * }}}
+   *
+   */
+  @inline def zipParRight[E, EE >: E, A, B](fa1: F[E, A], fa2: F[EE, B])(
+    implicit
+    CD: ConcurrentData2[F],
+    MD: Monoid[EE]
+  ): F[EE, B] =
+    zipWithPar(fa1, fa2)((_, b) => b)
+
+  /**
+   * Returns an effect that will acquire a resource and will release
+   * it after the execution of `use` regardless the fact that `use`
+   * succeed or fail.
+   *
+   * TODO: Example:
+   * {{{
+   *
+   * }}}
+   *
+   */
+  final def bracket[E, A, B](acquire: F[E, A], release: (A, Either[E, B]) => F[Nothing, Unit])(
+    use: A => F[E, B]
+  ): F[E, B] = {
+
+    implicit val _: Errorful2[F] = self
+
+    uninterruptible(
```

Comment:
Atop just `ensuring`, bracket will have a nasty implementation. Because the `use` action must be interruptible. You can look at the old ZIO code to see how to do this, it involves forking.

Reply:
Got it.